rfcs/0032-blockchain-in-hard-fork

```diff
@@ -0,0 +1,153 @@
+   - provide as a compile-time parameter
+
+- if we have a proof of the protocol state, provide that
+
+- subsequent blocks may use different proof, verification keys
+
+- force a software upgrade by bumping the protocol major version
+
+Notes
+-----
+
+- nodes will bootstrap from the new genesis state
+
+- time-locked accounts and transactions with expiration should continue
+   to work, because they rely on global slot numbers, and the
+   slot is incremented from the pause point
+
+Issues
+------
+
+- what is the source of randomness (epoch seed) for the post-fork chain?
+
+- is it acceptable to erase history (but we already have a succinct blockchain)
+
+- specifically, is there unfairness to the chains discarded past the pause point
+
+- what is allowed to change after the fork, with what effect:
+   - protocol state structure
+   - ledger structure/depth,
+   - account structure
+   - consensus constants
+
+- do we have to maintain pre-, post-fork code/data structures to verify the new genesis state,
+   ledger, and new blocks
+
+- if there's no proof of the ledger, how to convince the world that the new genesis ledger and
+   ledger are valid
+
+Alternative 2
+=============
+
+- choose a root and ledger at a pause point
+
+- nodes have a notion of "last fork time", an epoch-and-slot
+
+- upon a fork, gossip a special block to update the last fork time is issued, also referring to the
+   protocol state and proof at the pause point, and resetting the chain strength
+
+- if items affecting keys, ledger hashes, and proofs haven't changed, a software upgrade may not be required;
+   otherwise, force an upgrade
+
+Issues
+------
+
+- there may be empty slots or even empty epochs between the pause point and the restart; the reset
+   of the chain strength may be sufficient to handle the consequences
+
+- could we provide a new genesis block and proof, instead of the special block
+
+- As in Alternative (1), items may change after the fork, so there may be pre-, post-fork proofs, coda, data structures;
+   each additional fork may introduce more complexity
+
+- If the ledger/account structure changes, we won't have a proof for the post-fork state, also mentioned as
+   an issue for Alternative (1)
+
+- If the epoch, slot widths change, computations for epochs or slots may become more complex; subsequent forks may introduce
+   more complexity
+    - possible solution: a most-recent-fork timestamp (an earthly time, not the "last fork time") might be
+       used for epoch and slot calculations
+
+Alternative 3
+=============
+
+- as Alternative (2), except that a time offset is added to the special block
+   - eliminates the empty slot/epoch issue
+
+Issues
+-----
+
+- as in Alternative (2), except that the chain strength issue is resolved
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+There's no compelling reason not to do this. We need to be prepared to perform a hard fork if the main net
+fails.
+
```

Comment:
In short, I vote for Approach 3, with a slight modification. Below are the details.
Any approach here, I believe, should have the following two main goals:
- Goal 1: Retain the important desirable properties of our consensus protocol, such as bootstrappability from Genesis.
- Goal 2: Ensure fairness for every node. 

Downsides of Approach 1: 
A. It clearly lets go of the bootstrappability from Genesis property. In other words, what this approach does is like having a checkpoint that is enforced upon by a central authority, namely us. 
B. Imagine a section of the network that didn’t fail. There could be blocks generated by that network between the pause and restart times. It is not fair that those blocks were discarded at the restart. 
C. Regarding the part that we hardcode the last protocol state in the restart: The only part that we know the entire set of honest blocks have reached consensus upon is the block that is k blocks ago. Anything in the protocol state that comes after that time cannot be ensured to be a global common. So we cannot enforce the protocol state we have on the global set of nodes. 

Downsides of Approach 2: Chain quality is a fragile parameter in the sense that it depends on a few critical things. We would rather avoid having to do that if there is an approach that is as good. 

Approach 3: It doesn’t have any of the downsides of the other two approaches and achieves Goal 1. I believe we can also achieve Goal 2 in the following sense: the blocks that were generated after the pause time and before restart should just reappear after the restart with a (fair) caveat. Roughly speaking, the caveat is that if those blocks would not have been won if there weren’t a failure (because the failed nodes could have influenced the way the nodes that didn’t fail won), then they likely won’t. 
The exact way we can ensure this is to have the inputs to the vrf functions at restart time the same as they would have been at the slot following the failure if there were no failure. (So, slot number at restart = sl+1, where sl is the slot number at failure).

Reply:
thanks for the thoughtful explanation!